pkg/devfile/parser/data/2.0.0/components.go

```diff
@@ -126,47 +126,41 @@ func (d *Devfile200) UpdateComponent(component common.DevfileComponent) {
 }
 
 // GetCommands returns the slice of DevfileCommand objects parsed from the Devfile
-func (d *Devfile200) GetCommands() []common.DevfileCommand {
-	var commands []common.DevfileCommand
+func (d *Devfile200) GetCommands() map[string]common.DevfileCommand {
+	commands := make(map[string]common.DevfileCommand, len(d.Commands))
 
 	for _, command := range d.Commands {
 		// we convert devfile command id to lowercase so that we can handle
 		// cases efficiently without being error prone
 		// we also convert the odo push commands from build-command and run-command flags
-		if command.Exec != nil {
-			command.Exec.Id = strings.ToLower(command.Exec.Id)
-		} else if command.Composite != nil {
-			command.Composite.Id = strings.ToLower(command.Composite.Id)
-		}
+		commands[command.SetIDToLower()] = command
 
-		commands = append(commands, command)
 	}
 
 	return commands
 }
 
 // AddCommands adds the slice of DevfileCommand objects to the Devfile's commands
 // if a command is already defined, error out
-func (d *Devfile200) AddCommands(commands []common.DevfileCommand) error {
-	commandsMap := make(map[string]bool)
-	for _, command := range d.Commands {
-		commandsMap[command.Exec.Id] = true
-	}
+func (d *Devfile200) AddCommands(commands ...common.DevfileCommand) error {
+	commandsMap := d.GetCommands()
 
 	for _, command := range commands {
-		if _, ok := commandsMap[command.Exec.Id]; !ok {
+		id := command.GetID()
+		if _, ok := commandsMap[id]; !ok {
 			d.Commands = append(d.Commands, command)
 		} else {
-			return &common.AlreadyExistError{Name: command.Exec.Id, Field: "command"}
+			return &common.AlreadyExistError{Name: id, Field: "command"}
 		}
 	}
 	return nil
 }
 
 // UpdateCommand updates the command with the given id
 func (d *Devfile200) UpdateCommand(command common.DevfileCommand) {
+	id := strings.ToLower(command.Exec.Id)
```

Comment:
Why not call `command.GetID()` here? Otherwise, should we add a nil check for `command.Exec` and `command.Composite`?

Reply:
Agreed, I just ported the existing code, as-is, if I remember correctly.

Reply:
More precisely, the previous code used `command.Exec.Id` to identify which command to update. I guess this function doesn't take into account compound commands. I did wonder about it but I left the behavior unchanged because I'm just refactoring things here, not fixing bugs, though I could certainly change the behavior if we agree that's what we should be doing here. Wdyt, @johnmcollier @mik-dass ?

Reply:
Sure. This function was written before composite commands was implemented. So parent support won't work with composite commands. I will open a new issue for that.